haskell-src/Concordium/Utils/BinarySearch.hs

```diff
@@ -0,0 +1,63 @@
+-- |This module implements binary search on vectors.
+module Concordium.Utils.BinarySearch where
+
+import qualified Data.Vector as Vec
+
+-- |Perform a binary search on a sorted vector, returning the index and the value, if found.
+-- The first argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearchI :: (Ord k) => (a -> k) -> Vec.Vector a -> k -> Maybe (Int, a)
+binarySearchI valToKey vec key = search 0 (Vec.length vec - 1)
+  where
+    search lowIndex highIndex = case compare lowIndex highIndex of
+        LT ->
+            let midIndex = lowIndex + (highIndex - lowIndex) `div` 2
+                midVal = vec Vec.! midIndex
+             in case compare key (valToKey midVal) of
+                    LT -> search lowIndex (midIndex - 1)
+                    EQ -> Just (midIndex, midVal)
+                    GT -> search (midIndex + 1) highIndex
+        EQ ->
+            let val = vec Vec.! lowIndex
+             in if key == valToKey val
+                    then Just (lowIndex, val)
+                    else Nothing
+        GT -> Nothing
+
+-- |Perform a binary search on a sorted vector, returning the value, if found.
+-- The first argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearch :: (Ord k) => (a -> k) -> Vec.Vector a -> k -> Maybe a
+binarySearch valToKey vec key = snd <$> binarySearchI valToKey vec key
+
+-- |Perform a binary search on a sorted vector, returning the index and the value, if found.
+-- The first argument is a function that resolves an entry in a monad.
+-- The second argument is a function that maps a value to its search key.
+-- The elements of the vector must be sorted in ascending order with respect to the key, with no
+-- duplicate keys.
+binarySearchIM ::
+    (Ord k, Monad m) =>
+    (b -> m a) ->
+    (a -> k) ->
+    Vec.Vector b ->
+    k ->
+    m (Maybe (Int, a))
```

Comment:
This is not really entirely specified due to the monadic first argument. The order in which elements of the vector are accessed should be specified, and further preconditions stated, such as the key that is retrieved is always "equivalent" no matter the state of the computation.